### ISSUE TITLE
[Bug] RPC getBlockInfoByHeight "Returned block does not have a matching height"

### ISSUE THREAD

**Issue by KillianH** [casper-ecosystem/casper-js-sdk#103]
Hi !

I'm using the getBlockInfoByHeight from the casperServiceByJsonRPC but I'm getting this error on some blocks : 
"Returned block does not have a matching height"

It's not always and not fully reproductible I guess ...

I don't know if there's a cache mechanism with the client used to make the RPC calls ? (I think I've seen that somewhere in a changelog ...)

But I keep getting the error even if I loop over the same height.

Current block impacted on my side : 200003
Block returned : b6900e0b1967de9c527331f77b46970b1a4861a2c1c380c89f1b19c1a7bc8c0b / height : 200766

I'm not using a specific node because I've agregated some nodes behind a reverse proxy.

**Comment by hoffmannjan** [comment]
Hi @KillianH, I'll be back and able to check this on Monday, will let you know then. Thanks !

**Comment by KillianH** [comment]
Closing this. Didn't encountered this issue anymore. I managed to patch it on my own by getting the block by Hash. The issue occured when I tried to fetch multiple time the block I think.

**Comment by KillianH** [comment]
Sorry for the spam ... Actually the bug is still there and my fix for it doesn't work either.

Exemple for block 182515 :

```js
    try {
        return (await casperServiceByJsonRPC.getBlockInfoByHeight(height)).block;
    } catch (e) {
        const blockToFetch = await models.Block.findOne({
            where: {
                height: height,
            },
        });
        if (blockToFetch) {
            return (await casperServiceByJsonRPC.getBlockInfo(blockToFetch.hash)).block;
        }
        throw e;
    }
```

Errors : 

```js
Error: Returned block does not have a matching height.

Error: Tried to unwrap Error: Error: Invalid deploy: bodyHash missmatch. Expected: 209,154,210,201,34,167,214,149,54,97,240,231,219,136,9,55,87,7,42,189,56,173,230,46,222,71,201,108,81,151,160,202, 
                  got: 51,241,250,126,25,58,177,27,211,71,140,98,71,233,27,204,96,30,240,237,148,208,203,172,162,208,205,57,246,191,100,235.
```

**Comment by KillianH** [comment]
@hoffmannjan With the 2.5.1 I can't retrieve those deploy / blocks :

```
Missing 1 deploys out of 2 for block 228328
Missing 1 deploys out of 1 for block 228262
Missing 1 deploys out of 1 for block 228266
Missing 1 deploys out of 1 for block 235344

Failed to fetch deploy : fcf223f3053a53cee34e1fc7083664cd6f39db98ec95ac4861cb3f72de1e3233
Error: Tried to unwrap Error: Error: Invalid deploy: bodyHash missmatch. Expected: 193,148,75,27,20,17,42,7,246,95,86,189,126,142,89,111,224,35,28,5,111,143,166,165,188,232,85,111,76,31,126,151, 
                  got: 12,200,1,41,210,58,112,12,158,87,92,34,255,113,144,175,66,185,232,139,60,209,100,27,133,25,6,143,81,157,188,59.
    at Object.t.deployFromJson (/home/killian/WebstormProjects/CasperData/node_modules/casper-js-sdk/dist/lib.node.js:1:81076)
    at /home/killian/WebstormProjects/CasperData/node_modules/casper-js-sdk/dist/lib.node.js:1:58333
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at ErrImpl.unwrap (/home/killian/WebstormProjects/CasperData/node_modules/src/result.ts:140:15)
    at /home/killian/WebstormProjects/CasperData/node_modules/casper-js-sdk/dist/lib.node.js:1:58351
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
Failed to fetch deploy : 5fef6a04ffeef7f801feeac2cd7ebd4d5d13a1be4664ebd8609abdb33a1805a6
Error: Tried to unwrap Error: Error: Invalid deploy: bodyHash missmatch. Expected: 4,79,96,214,59,30,67,10,2,42,4,156,157,181,58,150,177,201,14,11,62,109,193,77,193,175,207,196,223,33,240,242, 
                  got: 126,231,178,201,184,110,198,162,78,106,145,30,148,75,151,73,157,105,213,226,145,37,211,39,146,82,219,15,2,168,102,18.
    at Object.t.deployFromJson (/home/killian/WebstormProjects/CasperData/node_modules/casper-js-sdk/dist/lib.node.js:1:81076)
    at /home/killian/WebstormProjects/CasperData/node_modules/casper-js-sdk/dist/lib.node.js:1:58333
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at ErrImpl.unwrap (/home/killian/WebstormProjects/CasperData/node_modules/src/result.ts:140:15)
    at /home/killian/WebstormProjects/CasperData/node_modules/casper-js-sdk/dist/lib.node.js:1:58351
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
Failed to fetch deploy : 14776981f3034faeaf314e0224641bfd264757f3a466aac27fda94bb88ec03bf
Error: Tried to unwrap Error: Error: Invalid deploy: bodyHash missmatch. Expected: 198,82,250,205,162,104,87,179,197,188,16,155,217,114,156,113,65,85,13,246,67,218,63,42,94,232,136,3,225,136,140,147, 
                  got: 182,46,181,5,182,235,75,190,192,43,218,44,249,23,233,128,13,112,158,31,122,13,165,66,203,175,246,60,102,215,8,184.
    at Object.t.deployFromJson (/home/killian/WebstormProjects/CasperData/node_modules/casper-js-sdk/dist/lib.node.js:1:81076)
    at /home/killian/WebstormProjects/CasperData/node_modules/casper-js-sdk/dist/lib.node.js:1:58333
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at ErrImpl.unwrap (/home/killian/WebstormProjects/CasperData/node_modules/src/result.ts:140:15)
    at /home/killian/WebstormProjects/CasperData/node_modules/casper-js-sdk/dist/lib.node.js:1:58351
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
Failed to fetch deploy : ffbff535506fd4945330ae2972d6cf2c538d041dab851f39e9b27897d0ee54f6
Error: Tried to unwrap Error: Error: Invalid deploy: bodyHash missmatch. Expected: 20,43,155,106,199,93,248,155,235,210,10,180,10,34,143,198,83,214,95,165,245,63,138,137,43,145,160,110,210,128,92,108, 
                  got: 145,152,127,53,4,224,93,52,162,151,24,203,244,204,137,166,210,152,126,38,4,61,207,123,152,159,90,243,77,75,189,231.
    at Object.t.deployFromJson (/home/killian/WebstormProjects/CasperData/node_modules/casper-js-sdk/dist/lib.node.js:1:81076)
    at /home/killian/WebstormProjects/CasperData/node_modules/casper-js-sdk/dist/lib.node.js:1:58333
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at ErrImpl.unwrap (/home/killian/WebstormProjects/CasperData/node_modules/src/result.ts:140:15)
    at /home/killian/WebstormProjects/CasperData/node_modules/casper-js-sdk/dist/lib.node.js:1:58351
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
```

With 2.5.2 I can parse them correctly. But with 2.5.2 I can't parse other deploys.

I'll add them in another comment in a few hours.  (Lunch time)

**Comment by hoffmannjan** [comment]
@KillianH you mean `2.5.1` is working correctly and `2.5.2` not, am I right?

Could you provide examples to reproduce this on our side?

Cheers

**Comment by KillianH** [comment]
> @KillianH you mean `2.5.1` is working correctly and `2.5.2` not, am I right?

I mean both versions have blocking issues on my side. I will send you the blocking deploys on 2.5.2 in a few hours here.

2.5.1 can't parse some deploys (but 2.5.2 can)
And 2.5.2 can't parse some deploys but 2.5.1 can

**Comment by hoffmannjan** [comment]
Interesting, as there shouldn't be any big changes between this verions. I'm really keen to check this. Thanks @KillianH

**Comment by KillianH** [comment]
```
Missing 1 deploys out of 2 for block 188890
Missing 1 deploys out of 1 for block 188899
Missing 1 deploys out of 1 for block 182515
Missing 1 deploys out of 1 for block 184637
Missing 1 deploys out of 2 for block 183638
Missing 1 deploys out of 2 for block 184603
Missing 3 deploys out of 6 for block 183661

Failed to retrieve transfer f390267f1ae2b87d22b23363ac54cec29958e00e86069890cd98a2c2bf069613
Error: Tried to unwrap Error: Error: Invalid deploy: bodyHash missmatch. Expected: 8,6,202,7,79,110,226,5,23,79,199,24,70,188,6,49,54,161,151,116,76,154,162,253,102,114,91,195,142,248,34,71, 
                  got: 160,72,234,16,145,80,100,203,216,198,105,131,255,210,24,249,85,147,36,157,222,226,236,127,243,212,255,60,17,215,139,23.
    at Object.t.deployFromJson (/home/killian/WebstormProjects/CasperData/node_modules/casper-js-sdk/dist/lib.node.js:1:81086)
    at /home/killian/WebstormProjects/CasperData/node_modules/casper-js-sdk/dist/lib.node.js:1:58343
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at ErrImpl.unwrap (/home/killian/WebstormProjects/CasperData/node_modules/src/result.ts:140:15)
    at /home/killian/WebstormProjects/CasperData/node_modules/casper-js-sdk/dist/lib.node.js:1:58361
    at processTicksAndRejections (node:internal/process/task_queues:96:5)

Failed to retrieve transfer 063079eb7e03717ef4d3fb4910a0ba16ac7ae0576ecc5177008ab930fd5911f6
Error: Tried to unwrap Error: Error: Invalid deploy: bodyHash missmatch. Expected: 145,107,159,177,173,180,7,7,237,60,52,239,255,184,73,74,90,8,189,74,240,148,94,70,88,54,183,18,79,78,100,56, 
                  got: 153,1,85,176,16,81,118,227,61,105,5,111,141,168,2,165,238,121,28,167,30,171,40,167,190,104,162,121,112,108,192,117.
    at Object.t.deployFromJson (/home/killian/WebstormProjects/CasperData/node_modules/casper-js-sdk/dist/lib.node.js:1:81086)
    at /home/killian/WebstormProjects/CasperData/node_modules/casper-js-sdk/dist/lib.node.js:1:58343
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at ErrImpl.unwrap (/home/killian/WebstormProjects/CasperData/node_modules/src/result.ts:140:15)
    at /home/killian/WebstormProjects/CasperData/node_modules/casper-js-sdk/dist/lib.node.js:1:58361
    at processTicksAndRejections (node:internal/process/task_queues:96:5)

Failed to retrieve transfer 3adbefd159f4fe30babaec26fa391f8fbbe20f695451bfe4125857b6b81610b3
Error: Tried to unwrap Error: Error: Invalid deploy: bodyHash missmatch. Expected: 90,37,84,146,246,251,227,96,244,226,61,234,93,57,197,62,18,90,239,28,163,218,166,231,225,28,113,168,74,92,64,237, 
                  got: 227,217,155,27,104,179,184,57,121,223,45,47,238,112,78,78,139,59,220,103,107,244,201,219,76,179,115,240,36,169,80,110.
    at Object.t.deployFromJson (/home/killian/WebstormProjects/CasperData/node_modules/casper-js-sdk/dist/lib.node.js:1:81086)
    at /home/killian/WebstormProjects/CasperData/node_modules/casper-js-sdk/dist/lib.node.js:1:58343
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at ErrImpl.unwrap (/home/killian/WebstormProjects/CasperData/node_modules/src/result.ts:140:15)
    at /home/killian/WebstormProjects/CasperData/node_modules/casper-js-sdk/dist/lib.node.js:1:58361
    at processTicksAndRejections (node:internal/process/task_queues:96:5)

Failed to retrieve transfer cff20de6e1ba61e4ba6c748468b73789450c8025641dea1f1c6daee4b2fc987a
Error: Tried to unwrap Error: Error: Invalid deploy: bodyHash missmatch. Expected: 209,154,210,201,34,167,214,149,54,97,240,231,219,136,9,55,87,7,42,189,56,173,230,46,222,71,201,108,81,151,160,202, 
                  got: 51,241,250,126,25,58,177,27,211,71,140,98,71,233,27,204,96,30,240,237,148,208,203,172,162,208,205,57,246,191,100,235.
    at Object.t.deployFromJson (/home/killian/WebstormProjects/CasperData/node_modules/casper-js-sdk/dist/lib.node.js:1:81086)
    at /home/killian/WebstormProjects/CasperData/node_modules/casper-js-sdk/dist/lib.node.js:1:58343
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at ErrImpl.unwrap (/home/killian/WebstormProjects/CasperData/node_modules/src/result.ts:140:15)
    at /home/killian/WebstormProjects/CasperData/node_modules/casper-js-sdk/dist/lib.node.js:1:58361
    at processTicksAndRejections (node:internal/process/task_queues:96:5)

Failed to retrieve transfer be93e58798032180da265b56995a10d9918834877b8dda793c16530394670c84
Error: Tried to unwrap Error: Error: Invalid deploy: bodyHash missmatch. Expected: 106,247,32,137,38,110,53,52,114,126,135,207,236,239,11,189,200,18,229,234,228,20,116,243,87,63,51,52,156,0,76,75, 
                  got: 65,102,38,227,242,213,84,29,108,131,244,148,67,104,222,142,15,10,4,45,101,116,143,66,14,121,148,49,72,198,9,177.
    at Object.t.deployFromJson (/home/killian/WebstormProjects/CasperData/node_modules/casper-js-sdk/dist/lib.node.js:1:81086)
    at /home/killian/WebstormProjects/CasperData/node_modules/casper-js-sdk/dist/lib.node.js:1:58343
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at ErrImpl.unwrap (/home/killian/WebstormProjects/CasperData/node_modules/src/result.ts:140:15)
    at /home/killian/WebstormProjects/CasperData/node_modules/casper-js-sdk/dist/lib.node.js:1:58361
    at processTicksAndRejections (node:internal/process/task_queues:96:5)

Failed to retrieve transfer f1926acce761139fd61ed95f9c5363acbea04593985fe85b0132b92cb50b5852
Error: Tried to unwrap Error: Error: Invalid deploy: bodyHash missmatch. Expected: 52,1,176,138,178,149,125,206,202,22,206,5,240,145,191,1,103,134,220,7,126,239,70,73,89,194,218,246,180,78,96,186, 
                  got: 141,127,185,208,192,241,54,23,152,192,153,123,232,9,111,134,31,92,101,145,66,241,229,4,170,142,250,66,61,158,120,71.
    at Object.t.deployFromJson (/home/killian/WebstormProjects/CasperData/node_modules/casper-js-sdk/dist/lib.node.js:1:81086)
    at /home/killian/WebstormProjects/CasperData/node_modules/casper-js-sdk/dist/lib.node.js:1:58343
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at ErrImpl.unwrap (/home/killian/WebstormProjects/CasperData/node_modules/src/result.ts:140:15)
    at /home/killian/WebstormProjects/CasperData/node_modules/casper-js-sdk/dist/lib.node.js:1:58361
    at processTicksAndRejections (node:internal/process/task_queues:96:5)

Failed to retrieve transfer 949601d446e484c8c7097679c8ace642ade9e21690861d0bbac77d8ea066d531
Error: Tried to unwrap Error: Error: Invalid deploy: bodyHash missmatch. Expected: 145,107,159,177,173,180,7,7,237,60,52,239,255,184,73,74,90,8,189,74,240,148,94,70,88,54,183,18,79,78,100,56, 
                  got: 153,1,85,176,16,81,118,227,61,105,5,111,141,168,2,165,238,121,28,167,30,171,40,167,190,104,162,121,112,108,192,117.
    at Object.t.deployFromJson (/home/killian/WebstormProjects/CasperData/node_modules/casper-js-sdk/dist/lib.node.js:1:81086)
    at /home/killian/WebstormProjects/CasperData/node_modules/casper-js-sdk/dist/lib.node.js:1:58343
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at ErrImpl.unwrap (/home/killian/WebstormProjects/CasperData/node_modules/src/result.ts:140:15)
    at /home/killian/WebstormProjects/CasperData/node_modules/casper-js-sdk/dist/lib.node.js:1:58361
    at processTicksAndRejections (node:internal/process/task_queues:96:5)

Failed to retrieve transfer 509dd9c766a8e43330f449a7953d95a0caf02a18dc9d9dce511ffc2355587016
Error: Tried to unwrap Error: Error: Invalid deploy: bodyHash missmatch. Expected: 189,127,207,132,162,37,176,64,173,160,73,32,85,55,255,35,236,75,253,210,126,26,50,152,255,212,60,68,38,23,122,71, 
                  got: 157,28,26,170,201,109,255,187,44,248,96,170,122,248,157,132,195,29,89,187,150,234,15,31,71,225,79,126,191,27,158,168.
    at Object.t.deployFromJson (/home/killian/WebstormProjects/CasperData/node_modules/casper-js-sdk/dist/lib.node.js:1:81086)
    at /home/killian/WebstormProjects/CasperData/node_modules/casper-js-sdk/dist/lib.node.js:1:58343
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at ErrImpl.unwrap (/home/killian/WebstormProjects/CasperData/node_modules/src/result.ts:140:15)
    at /home/killian/WebstormProjects/CasperData/node_modules/casper-js-sdk/dist/lib.node.js:1:58361
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    
Failed to retrieve transfer 4816c77c57cab8df2879277e02108aa11f6ae4cbe8188cd1d17eadc1b89819a6
Error: Tried to unwrap Error: Error: Invalid deploy: bodyHash missmatch. Expected: 159,233,246,222,92,246,151,202,13,69,230,149,228,128,227,252,37,11,11,136,227,208,41,63,181,18,43,5,230,254,79,115, 
                  got: 41,70,150,194,31,2,144,203,92,44,50,63,115,125,70,177,91,70,180,34,224,157,2,218,180,80,118,148,211,67,53,158.
    at Object.t.deployFromJson (/home/killian/WebstormProjects/CasperData/node_modules/casper-js-sdk/dist/lib.node.js:1:81086)
    at /home/killian/WebstormProjects/CasperData/node_modules/casper-js-sdk/dist/lib.node.js:1:58343
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at ErrImpl.unwrap (/home/killian/WebstormProjects/CasperData/node_modules/src/result.ts:140:15)
    at /home/killian/WebstormProjects/CasperData/node_modules/casper-js-sdk/dist/lib.node.js:1:58361
    at processTicksAndRejections (node:internal/process/task_queues:96:5)

```

Those are the transfers I can't parse with 2.5.2 & 2.6.0. But I can parse them correctly with 2.5.1

**Comment by KillianH** [comment]
```ts
import { CasperClient } from "casper-js-sdk";

const casperClient = new CasperClient("http://3.14.161.135:7777/rpc")

async function test() {
    const deployArrayBug = [
        "f390267f1ae2b87d22b23363ac54cec29958e00e86069890cd98a2c2bf069613",
        "063079eb7e03717ef4d3fb4910a0ba16ac7ae0576ecc5177008ab930fd5911f6",
        "3adbefd159f4fe30babaec26fa391f8fbbe20f695451bfe4125857b6b81610b3",
        "cff20de6e1ba61e4ba6c748468b73789450c8025641dea1f1c6daee4b2fc987a",
        "be93e58798032180da265b56995a10d9918834877b8dda793c16530394670c84",
        "f1926acce761139fd61ed95f9c5363acbea04593985fe85b0132b92cb50b5852",
        "949601d446e484c8c7097679c8ace642ade9e21690861d0bbac77d8ea066d531",
        "509dd9c766a8e43330f449a7953d95a0caf02a18dc9d9dce511ffc2355587016",
        "4816c77c57cab8df2879277e02108aa11f6ae4cbe8188cd1d17eadc1b89819a6",
        "fcf223f3053a53cee34e1fc7083664cd6f39db98ec95ac4861cb3f72de1e3233",
        "5fef6a04ffeef7f801feeac2cd7ebd4d5d13a1be4664ebd8609abdb33a1805a6",
        "14776981f3034faeaf314e0224641bfd264757f3a466aac27fda94bb88ec03bf",
        "ffbff535506fd4945330ae2972d6cf2c538d041dab851f39e9b27897d0ee54f6",
    ]

    for (const hash of deployArrayBug) {
        try {
            const deploy = await casperClient.getDeploy(hash)
            console.log("Deploy : " + deploy[1].deploy.hash + " OK");
        } catch (e) {
            console.log("Deploy : " + hash + " KO");
            console.log(e);
        }
    }
}

test();
```

It's reproductible with this code if you switch between 2.5.1 and 2.5.2/2.6.0

Mainnet read only node used for this example.

To recap : 

2.0.0 -> Working fine :exploding_head: 

2.0.1 to 2.5.1 -> Last 4 deploys KO

Deploy : fcf223f3053a53cee34e1fc7083664cd6f39db98ec95ac4861cb3f72de1e3233 KO
Deploy : 5fef6a04ffeef7f801feeac2cd7ebd4d5d13a1be4664ebd8609abdb33a1805a6 KO
Deploy : 14776981f3034faeaf314e0224641bfd264757f3a466aac27fda94bb88ec03bf KO
Deploy : ffbff535506fd4945330ae2972d6cf2c538d041dab851f39e9b27897d0ee54f6 KO

Those did include "0" value in the payment or session so it's normal that's fixed with 2.5.2.

2.5.2 to 2.6.0 -> First 9 deploys KO

Deploy : f390267f1ae2b87d22b23363ac54cec29958e00e86069890cd98a2c2bf069613 KO
Deploy : 063079eb7e03717ef4d3fb4910a0ba16ac7ae0576ecc5177008ab930fd5911f6 KO
Deploy : 3adbefd159f4fe30babaec26fa391f8fbbe20f695451bfe4125857b6b81610b3 KO
Deploy : cff20de6e1ba61e4ba6c748468b73789450c8025641dea1f1c6daee4b2fc987a KO
Deploy : be93e58798032180da265b56995a10d9918834877b8dda793c16530394670c84 KO
Deploy : f1926acce761139fd61ed95f9c5363acbea04593985fe85b0132b92cb50b5852 KO
Deploy : 949601d446e484c8c7097679c8ace642ade9e21690861d0bbac77d8ea066d531 KO
Deploy : 509dd9c766a8e43330f449a7953d95a0caf02a18dc9d9dce511ffc2355587016 KO
Deploy : 4816c77c57cab8df2879277e02108aa11f6ae4cbe8188cd1d17eadc1b89819a6 KO

Didn't digged into the code but seems like 2.0.1 introduce a bug 2.5.2 fixed it (u512 0 value) but introduced another one.

All deploy here are transfers if I'm correct. Some are correctly deployed on the blockchain and other are failed. At first glance I don't see any weird value in the deploys but you may find something that I don't see myself.

**Comment by hoffmannjan** [comment]
@KillianH thank you for letting us know, the bug were here for some time. In fact, it is also here in 2.0.0, if you log out `deploy` in your `try` code, you will see that there wasn't any error thrown (just `undefined`). That was the only change in `2.0.1`. 

Will let you know when this wil be fixed, thanks again!

**Comment by KillianH** [comment]
Okay great 👍 Feel free to ping me anytime :) I can test the solution over the whole mainnet under 20min or so.

**Comment by hoffmannjan** [comment]
@KillianH please check this fix https://github.com/casper-ecosystem/casper-js-sdk/pull/115 :) 
It should be released soon.

**Comment by hoffmannjan** [comment]
Fix is released in `2.6.1`, so we can close the issue. Thank you @KillianH !
